File: src/03-hard/roller-coaster/solution.py
L, C, N = map(int, input().split())
groups = [int(input()) for _ in range(N)]

total = 0
rides = 0
head = 0
memo = {}

while rides < C:
    if head in memo:
        prev_rides, prev_total = memo[head]
        cycle_len = rides - prev_rides
        cycle_earn = total - prev_total
        remaining = C - rides
        full_cycles = remaining // cycle_len
        total += full_cycles * cycle_earn
        rides += full_cycles * cycle_len
        break
    memo[head] = (rides, total)

    cap = L
    earned = 0
    idx = head
    for _ in range(N):
        if groups[idx] <= cap:
            cap -= groups[idx]
            earned += groups[idx]
            idx = (idx + 1) % N
        else:
            break
    total += earned
    rides += 1
    head = idx

# Simulate remaining rides after cycle skip
while rides < C:
    cap = L
    earned = 0
    idx = head
    for _ in range(N):
        if groups[idx] <= cap:
            cap -= groups[idx]
            earned += groups[idx]
            idx = (idx + 1) % N
        else:
            break
    total += earned
    rides += 1
    head = idx

print(total)

File: src/03-hard/roller-coaster/solution.clj
(ns Solution
  (:gen-class))

(defn -main [& args]
  (let [[L C N] (map #(Long/parseLong %) (.split (read-line) " "))
        groups (vec (repeatedly N #(Long/parseLong (read-line))))
        ;; For each starting head position, compute (next-head, earned)
        sim (fn [head]
              (loop [cap L earned 0 idx head cnt 0]
                (if (or (>= cnt N) (> (groups idx) cap))
                  [idx earned]
                  (recur (- cap (groups idx))
                         (+ earned (groups idx))
                         (mod (inc idx) N)
                         (inc cnt)))))
        ;; Precompute transitions for all N positions
        transitions (vec (map sim (range N)))]
    (loop [rides 0 head 0 total 0 memo {}]
      (if (>= rides C)
        (println total)
        (if (contains? memo head)
          (let [[prev-rides prev-total] (memo head)
                cycle-len (- rides prev-rides)
                cycle-earn (- total prev-total)
                remaining (- C rides)
                full-cycles (quot remaining cycle-len)
                new-total (+ total (* full-cycles cycle-earn))
                new-rides (+ rides (* full-cycles cycle-len))]
            ;; Simulate remaining
            (loop [r new-rides h head t new-total]
              (if (>= r C)
                (println t)
                (let [[nh earned] (transitions h)]
                  (recur (inc r) nh (+ t earned))))))
          (let [[nh earned] (transitions head)]
            (recur (inc rides) nh (+ total earned)
                   (assoc memo head [rides total]))))))))
